### ISSUE TITLE
execute?

### ISSUE THREAD

**Issue by wboeke** [CCurl/c3#51]
Is it very hard to add the forth word 'execute' to c3? Then interesting things could be done, e.g.:

2 3 ' + execute .

Or, more useful:
variable #msg
create values  ' @  , ' !  ,
: value_ ( n 'name' )
   create ,
   does>  values #msg @ cells + @ execute  0 #msg ! ;
: to_  1 #msg ! ;
0 value_ vv vv . cr
10 to_ vv vv . cr

**Comment by CCurl** [comment]
Thre is already a word for that ... EXEC.

But this should do it if you want the EXECUTE ...

: execute >R ;

FWIW, ' pushes 3 things onto the stack ... if found ( -- xt flags 1), else just ( -- 0)

So, 

1 2 ' + drop drop execute
![image](https://github.com/CCurl/c3/assets/14334380/ccae78da-1f5b-4076-947e-4f48dfe0b86e)

**Comment by wboeke** [comment]
Sorry, but there is no 'exec' in c3, there is an 'execute' in c4. At that, your solution does not work in my last (more interesting) example. In my opinion it indeed cannot work, it is too simple.

**Comment by CCurl** [comment]
Um, are you on current code? The image I posted above is definately from the c3 I am running (the main in github).

![image](https://github.com/CCurl/c3/assets/14334380/0d23e5af-ea97-4848-a070-f29b078d0c50)

**Comment by CCurl** [comment]
![image](https://github.com/CCurl/c3/assets/14334380/f688153c-63e9-4b3b-be22-efe2fb520534)

**Comment by CCurl** [comment]
![image](https://github.com/CCurl/c3/assets/14334380/fb5782b3-e249-4005-96cc-ed157872e2cb)

**Comment by wboeke** [comment]
You are right, 'exec' is present. However, my main issue stil exists. For clarity, I repeat it here:

variable #msg
create values ' @ , ' ! ,
: value_ ( n 'name' )
  create ,
  does> values #msg @ cells + @ exec 0 #msg ! ;
: to_ 1 #msg ! ;
0 value_ vv vv . cr
10 to_ vv vv . cr

The result should be: 0 10
but alas, it is different.

**Comment by CCurl** [comment]
Try this, it appears to do what you want

variable #msg 0 #msg !
create values (EXIT) C,
' @ 2DROP V, ' ! 2DROP V,
: value_ ( n 'name' )
    create  does> 
    values #msg @ cells + @ exec 0 #msg ! ;
: to_ 1 #msg ! ;
0 value_ vv vv . cr
10 to_ vv vv . cr

**Comment by wboeke** [comment]
Bingo! It works!
Now going to find out _how_ it works.

**Comment by CCurl** [comment]
Note that CREATE only generates code to push VHERE, but it does not "complete the word" if you will ... meaning that it does NOT generate an EXIT. It is intended to be used with DOES>, or to be completed manually so you can generate whatever code you want, in which case you have to manually include the EXIT at the end.

So CREATE var without the (EXIT) C, will be a problem.

Also note that it is "V," not "," ... as , puts the value in the CODE memory space, while V, puts the value in the VARS memory space, while is what CREATE generates; a address in the VARS memory space.

Lastly, ' pushes 3 values when found, (-- XT flags 1) so you have to discard the top 2 items to get to the XT. When not found, ' only pushes 1 value ( -- 0).

I hope that helps a little. :)

**Comment by CCurl** [comment]
I use these instead of VALUE / TO
```
: val    CREATE 0 V, 10 C, (EXIT) C, ; \ Opcode #10 is FETCH (@)
: >val   VHERE CELL - CONSTANT (HERE) -- 8 C, (EXIT) C, ; \ Opcode #8 is STORE (!)
: (val)  VHERE CELL - CONSTANT ;
\ Here is a definition of >val that uses DOES>
\ it might be more "elegant", but the generated code is longer and less efficient
\ : >val  CREATE DOES> CELL - ! ;
```
I use them like this:
```
val src    \ ( -- ) VALUE src
>val >src  \ ( N-- ) TO-src, similar to the way >R is TO-R

src . ( prints 0 )
3345 >src
src . ( prints 3345 )
```

**Comment by wboeke** [comment]
Yes, I knew. My point was about the forth words 'value' and 'to', how to implement this.
Thanks for your explanation, you may close this issue if you want.
